File: packages/splunk_otel_flutter/splunk_otel_flutter/android/src/main/kotlin/com/splunk/rum/flutter/extensions/StatusExtension.kt
/*
 * Copyright 2025 Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.rum.flutter.extensions

import com.splunk.rum.flutter.GeneratedStatus
import com.splunk.rum.integration.agent.api.Status

fun GeneratedStatus.toAgentApiStatus(): Status {
    return when (this) {
        GeneratedStatus.RUNNING -> Status.Running
        GeneratedStatus.NOT_INSTALLED -> Status.NotRunning.NotInstalled
        GeneratedStatus.SUB_PROCESS -> Status.NotRunning.Subprocess
        GeneratedStatus.SAMPLED_OUT -> Status.NotRunning.SampledOut
        GeneratedStatus.UNSUPPORTED_OS_VERSION -> Status.NotRunning.UnsupportedOsVersion
        GeneratedStatus.UNSUPPORTED_PLATFORM -> throw IllegalArgumentException("Unsupported GeneratedStatus type for UNSUPPORTED_PLATFORM")
    }
}

fun Status.toGeneratedStatus(): GeneratedStatus {
    return when (this) {
        Status.Running -> GeneratedStatus.RUNNING
        Status.NotRunning.NotInstalled -> GeneratedStatus.NOT_INSTALLED
        Status.NotRunning.SampledOut -> GeneratedStatus.SAMPLED_OUT
        Status.NotRunning.Subprocess -> GeneratedStatus.SUB_PROCESS
        Status.NotRunning.UnsupportedOsVersion -> GeneratedStatus.UNSUPPORTED_OS_VERSION
    }
}

File: packages/splunk_otel_flutter/splunk_otel_flutter/ios/splunk_otel_flutter/Sources/splunk_otel_flutter/extensions/StatusExtension.swift
//
/*
Copyright 2025 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import SplunkAgent

enum StatusConversionError: Error {
    case invalidStatus(String)
}

extension GeneratedStatus {
    func toAgentApiStatus() throws -> Status {
        switch self {
        case .running:
            return .running
        case .notInstalled:
            return .notRunning(.notInstalled)
        case .sampledOut:
            return .notRunning(.sampledOut)
        case .unsupportedOsVersion:
            return .notRunning(.unsupportedPlatform)
        case .subprocess:
            throw StatusConversionError.invalidStatus("invalid status subprocess")
        case .unsupportedPlatform:
            throw StatusConversionError.invalidStatus("invalid status unsupportedPlatform")
        }
    }
}

extension Status {
    func toGeneratedStatus() -> GeneratedStatus {
        switch self {
        case .running: return .running
        case .notRunning(.notInstalled): return .notInstalled
        case .notRunning(.sampledOut): return .sampledOut
        case .notRunning(.unsupportedPlatform): return .unsupportedPlatform
        default: return .unsupportedOsVersion
        }
    }
}
